bundle/permissions/utils.go

```diff
@@ -0,0 +1,86 @@
+package permissions
+
+import (
+	"context"
+
+	"github.com/databricks/cli/bundle/config/resources"
+	"github.com/databricks/cli/libs/cmdio"
+	"github.com/databricks/cli/libs/diag"
+)
+
+func convert(
+	ctx context.Context,
+	topLevelPermissions []resources.Permission,
+	resourcePermissions []resources.Permission,
+	resourceName string,
+	lm map[string]string,
+) []resources.Permission {
+	permissions := make([]resources.Permission, 0)
+	for _, p := range topLevelPermissions {
+		level, ok := lm[p.Level]
+		// If there is no top level permission level defined in the map, it means
+		// it's not applicable for the resource, therefore skipping
+		if !ok {
+			continue
+		}
+
+		if notifyForPermissionOverlap(ctx, p, resourcePermissions, resourceName) {
+			continue
+		}
+
+		permissions = append(permissions, resources.Permission{
+			Level:                level,
+			UserName:             p.UserName,
+			GroupName:            p.GroupName,
+			ServicePrincipalName: p.ServicePrincipalName,
+		})
+	}
+
+	return permissions
+}
+
+func isPermissionOverlap(
+	permission resources.Permission,
+	resourcePermissions []resources.Permission,
+	resourceName string,
+) (bool, diag.Diagnostics) {
+	diagnostics := make(diag.Diagnostics, 0)
+	for _, rp := range resourcePermissions {
+		if rp.GroupName != "" && rp.GroupName == permission.GroupName {
+			diagnostics = diagnostics.Extend(
+				diag.Warningf("'%s' already has permissions set for '%s' group", resourceName, rp.GroupName),
+			)
+		}
+
+		if rp.UserName != "" && rp.UserName == permission.UserName {
+			diagnostics = diagnostics.Extend(
+				diag.Warningf("'%s' already has permissions set for '%s' user name", resourceName, rp.UserName),
+			)
+		}
+
+		if rp.ServicePrincipalName != "" && rp.ServicePrincipalName == permission.ServicePrincipalName {
+			diagnostics = diagnostics.Extend(
+				diag.Warningf("'%s' already has permissions set for '%s' service principal name", resourceName, rp.ServicePrincipalName),
+			)
+		}
+	}
+
+	return len(diagnostics) > 0, diagnostics
+}
+
+func notifyForPermissionOverlap(
+	ctx context.Context,
+	permission resources.Permission,
+	resourcePermissions []resources.Permission,
+	resourceName string,
+) bool {
+	isOverlap, diagnostics := isPermissionOverlap(permission, resourcePermissions, resourceName)
+	// If there is permission overlap, show a warning to the user
+	if isOverlap {
+		for _, d := range diagnostics {
+			cmdio.LogString(ctx, d.Summary)
```

Comment:
Can you leave this one out for now? I'd like for us to collect all diagnostics at the top level and visualize jointly.